From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sun, 3 Aug 2025 20:38:56 -0500
Subject: [PATCH] Fish World Schedulers

This patch adds a ReentrantReadWriteLock, RegionizedWorldData and Folia's Entity Scheduler.
RegionizedWorldData bundles an EntityScheduler.EntitySchedulerTickList and a WorldRegionScheduler.Scheduler
  effectively implementing PWT-compatible versions of Folia's Entity and Location schedulers, instead
  of using Paper's fallback schedulers, which run on the main thread. This means the task load is now
  effectively spread across worlds.
ReentrantReadWriteLock allows consistent read-write blocking logic for scheduling async tasks into world's
  threads as a compatibility layer for Bukkit plugins. The write requests will be scheduled at the end
  of the world tick if the world is currently ticking, or will run immediately and temporarily block the
  read lock if not. This scheduling logic is very performant as long as there're no plugins spamming
  CraftWorld or CraftBlock requests.
The Folia Entity Scheduler allows running tasks in whatever thread the region that owns the entity is ticking.
  In Paper, the Entity Scheduler falls back to executing in the main thread before entity ticking, effectively
  making it compatible with PWT. This has been improved by moving the entity scheduler into the world thread,
  keeping the compatibility, but allowing to execute these tasks in its corresponding threads, leveraging most
  of the load off the main thread. This also allows plugins to support PWT without explicit API.

Additionally, PacketUtil requires validating against the current ticking thread (world or main thread), instead
  of only the main thread. This prevents issues when scheduling tasks into the world thread like opening
  an inventory while another is currently opened in a scheduled task.

diff --git a/net/minecraft/network/protocol/PacketUtils.java b/net/minecraft/network/protocol/PacketUtils.java
index 144c18c08bf8f03d907b97f968aa171a9fbf997f..ac531086508f38d6af31670968d6ae710a2aea1c 100644
--- a/net/minecraft/network/protocol/PacketUtils.java
+++ b/net/minecraft/network/protocol/PacketUtils.java
@@ -15,6 +15,15 @@ public class PacketUtils {
     private static final Logger LOGGER = LogUtils.getLogger();
 
     public static <T extends PacketListener> void ensureRunningOnSameThread(final Packet<T> packet, final T listener, final ServerLevel level) throws RunningOnDifferentThreadException {
+        // Fish start - Location scheduler handling
+        if (me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled) {
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(level)) {
+                level.getServer().packetProcessor().scheduleIfPossible(listener, packet);
+                throw RunningOnDifferentThreadException.RUNNING_ON_DIFFERENT_THREAD;
+            }
+            return;
+        }
+        // Fish start - Location scheduler handling
         ensureRunningOnSameThread(packet, listener, level.getServer().packetProcessor());
     }
 
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index ae49363368d1d6d22611d7a075a54d8b238219c7..68c6a91b08b5c593eb0a822157bbb909288504bb 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -336,6 +336,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     private final Set<String> pluginsBlockingSleep = new java.util.HashSet<>(); // Paper - API to allow/disallow tick sleeping
     public static final long SERVER_INIT = System.nanoTime(); // Paper - Lag compensation
     public gg.pufferfish.pufferfish.util.AsyncExecutor pufferfish$mobSpawnExecutor = new gg.pufferfish.pufferfish.util.AsyncExecutor("Pufferfish Async Mob Spawning Thread"); // Pufferfish - Optimize mob spawning
+    public final me.biquaternions.fish.threadedregions.FallbackWorldData fish$globalData = new me.biquaternions.fish.threadedregions.FallbackWorldData(); // Fish - Global Region
     // Paper start - improve tick loop
     public final ca.spottedleaf.common.time.TickData tickTimes1s  = new ca.spottedleaf.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(1L));
     public final ca.spottedleaf.common.time.TickData tickTimes5s  = new ca.spottedleaf.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(5L));
@@ -1770,6 +1771,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.server.getScheduler().mainThreadHeartbeat(); // CraftBukkit
         // Paper start - optimise Folia entity scheduler
         ((io.papermc.paper.threadedregions.scheduler.FoliaGlobalRegionScheduler)org.bukkit.Bukkit.getGlobalRegionScheduler()).tick();
+        if (!me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled) { // Fish - Handle Entity Schedulers
         for (io.papermc.paper.threadedregions.EntityScheduler scheduler : this.entitySchedulerTickList.getAllSchedulers()) {
             if (scheduler.isRetired()) {
                 continue;
@@ -1777,6 +1779,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
             scheduler.executeTick();
         }
+        } // Fish - Handle Entity Schedulers
         // Paper end - optimise Folia entity scheduler
         // Paper end - Folia scheduler API
         io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.ADVENTURE_CLICK_MANAGER.handleQueue(this.tickCount); // Paper
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 05d9b505245729c2939bbe4832e891434c62dc23..c90cc7a9c82a715a83700ca7f733bff64335277f 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -238,6 +238,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private static final org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry();
     public final org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer persistentDataContainer = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(DATA_TYPE_REGISTRY);
     private final alternate.current.wire.WireHandler wireHandler = new alternate.current.wire.WireHandler(this); // Paper - optimize redstone (Alternate Current)
+    public final java.util.concurrent.locks.ReentrantReadWriteLock fish$lock = new java.util.concurrent.locks.ReentrantReadWriteLock(true); // Fish - Schedule tasks for end of tick
+    public final me.biquaternions.fish.threadedregions.RegionizedWorldData fish$worldData = new me.biquaternions.fish.threadedregions.RegionizedWorldData(); // Fish - World Data
 
     @Override
     public @Nullable LevelChunk getChunkIfLoaded(int x, int z) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 7f7e179ab67cc1f7439e963b6c06a2328a0acb85..db3be0c67ab88ebdcec51ff4b92a75bc9d2d1d68 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -5243,6 +5243,10 @@ public abstract class Entity
     // Paper end - Folia schedulers
     // Paper start - optimise Folia entity scheduler
     public final void registerScheduler() {
+        if (me.biquaternions.fish.FishConfig.getInstance().async.worldTicking.enabled && this.level instanceof ServerLevel serverLevel) {
+            this.getBukkitEntity().taskScheduler.registerTo(serverLevel.fish$worldData.entitySchedulerTickList);
+            return;
+        }
         this.getBukkitEntity().taskScheduler.registerTo(net.minecraft.server.MinecraftServer.getServer().entitySchedulerTickList);
     }
     // Paper end - optimise Folia entity scheduler
